crates/pg_workspace_new/src/workspace/server/change.rs

```diff
@@ -54,235 +63,278 @@ impl Document {
-        // e.g. if a statement is "select 1", and the change is "select 2; select 2", its an addition even though its in the middle of the statement.
-        // hence we only have three "real" cases:
-        // 1. the change touches no statement at all (addition)
-        // 2. the change touches exactly one statement AND splitting the statement results in just
-        //    one statement (modification)
-        // 3. the change touches more than one statement (addition/deletion)
+        changes
+    }
 
-        let new_content = change.apply_to_text(&self.content);
+    fn insert_statement(&mut self, range: TextRange) -> usize {
+        let pos = self
+            .positions
+            .binary_search_by(|(_, r)| r.start().cmp(&range.start()))
+            .unwrap_err();
 
-        let mut affected = vec![];
+        let new_id = self.id_generator.next();
+        self.positions.insert(pos, (new_id, range));
 
-        for (idx, (id, r)) in self.statements.iter_mut().enumerate() {
-            if r.intersect(change.range.unwrap()).is_some() {
-                affected.push((idx, (*id, *r)));
-            } else if r.start() > change.range.unwrap().end() {
-                if change.is_addition() {
-                    *r += change.diff_size();
-                } else if change.is_deletion() {
-                    *r -= change.diff_size();
-                }
+        new_id
+    }
+
+    /// Returns all relevant details about the change and its effects on the current state of the document.
+    /// - The affected range is the full range of the change, including the range of all statements that intersect with the change
+    /// - All indices of affected statement positions
+    /// - The index of the first statement position before the change, if any
+    /// - The index of the first statement position after the change, if any
+    /// - the full affected range includng the prev and next statement
+    fn get_affected(
+        &self,
+        change_range: TextRange,
+        content_size: TextSize,
+        diff_size: TextSize,
+        is_addition: bool,
+    ) -> Affected {
+        let mut start = change_range.start();
+        let mut end = change_range.end().min(content_size);
+
+        let mut affected_indices = Vec::new();
+        let mut prev_index = None;
+        let mut next_index = None;
+
+        for (index, (_, pos_range)) in self.positions.iter().enumerate() {
+            if pos_range.intersect(change_range).is_some() {
+                affected_indices.push(index);
+                start = start.min(pos_range.start());
+                end = end.max(pos_range.end());
+            } else if pos_range.end() <= change_range.start() {
+                prev_index = Some(index);
+            } else if pos_range.start() >= change_range.end() && next_index.is_none() {
+                next_index = Some(index);
+                break;
```

Comment:
nit: we're implicitly assuming that the `StatementPosition`s are sorted ascending by range. 

If we change the order that the `StatementSplitter` returns statements, this will break.

Should we make the ordering more explicit?

Reply:
good point. I added a test to the statement splitter to ensure the output is always sorted. is that sufficient? I think it's easier to make sure it there than to handle it everywhere.

Reply:
yeah, great solution 👍